core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -358,7 +358,7 @@ class LogSegment private[log] (val log: FileRecords,
 
         if (batch.magic >= RecordBatch.MAGIC_VALUE_V2) {
           leaderEpochCache.foreach { cache =>
-            if (batch.partitionLeaderEpoch > cache.latestEpoch) // this is to avoid unnecessary warning in cache.assign()
+            if (cache.latestEpoch.forall(batch.partitionLeaderEpoch > _))
```

Comment:
As I can trace back, it is not equivalent with the old code as in the case when -1 > -1 it would give `false`, however now `forall` returns `true` for `None` (which was -1 in the old code).
So if there are no latest epoch in the cache we return true regardless of what's in the batch. Is this desired or a case which couldn't happen?

Reply:
Hard to tell whether the current behavior was intentional. I guess the question is how we should handle an invalid epoch in the batch. The current code silently ignores the batch while this change would insert it into the cache if there were no entries. Batches like that shouldn't exist of course, so we could raise an exception, but I think I will just alter the check and make the current behavior explicit.